arch/riscv/core/pmp/core_pmp.c

```diff
@@ -0,0 +1,605 @@
+		csr_write(0x3B8, value); break;
+	case CSR_PMPADDR9:
+		csr_write(0x3B9, value); break;
+	case CSR_PMPADDR10:
+		csr_write(0x3BA, value); break;
+	case CSR_PMPADDR11:
+		csr_write(0x3BB, value); break;
+	case CSR_PMPADDR12:
+		csr_write(0x3BC, value); break;
+	case CSR_PMPADDR13:
+		csr_write(0x3BD, value); break;
+	case CSR_PMPADDR14:
+		csr_write(0x3BE, value); break;
+	case CSR_PMPADDR15:
+		csr_write(0x3BF, value); break;
+	default:
+		break;
+	}
+}
+
+int z_riscv_pmp_set(unsigned int index, ulong_t cfg_val, ulong_t addr_val)
+{
+	ulong_t reg_val;
+	ulong_t shift, mask;
+	int pmpcfg_csr;
+	int pmpaddr_csr;
+
+	if ((index >= PMP_SLOT_NUMBER) | (index < 0))
+		return -1;
+
+	/* Calculate PMP config/addr register, shift and mask */
+#ifdef CONFIG_64BIT
+	pmpcfg_csr = CSR_PMPCFG0 + (index >> 4);
+	shift = (index & 0x7) << 3;
+#else
+	pmpcfg_csr = CSR_PMPCFG0 + (index >> 2);
+	shift = (index & 0x3) << 3;
+#endif /* CONFIG_64BIT */
+	pmpaddr_csr = CSR_PMPADDR0 + index;
+
+	/* Mask = 0x000000FF<<((index%4)*8) */
+	mask = 0x000000FF << shift;
+
+	cfg_val = cfg_val << shift;
+	addr_val = addr_val >> PMP_SHIFT_ADDR;
+
+	reg_val = csr_read_enum(pmpcfg_csr);
+	reg_val = reg_val & ~mask;
+	reg_val = reg_val | cfg_val;
+
+	csr_write_enum(pmpaddr_csr, addr_val);
+	csr_write_enum(pmpcfg_csr, reg_val);
+
+	return 0;
+}
+
+int pmp_get(unsigned int index, ulong_t *cfg_val, ulong_t *addr_val)
+{
+	ulong_t shift;
+	int pmpcfg_csr;
+	int pmpaddr_csr;
+
+	if ((index >= PMP_SLOT_NUMBER) | (index < 0))
+		return -1;
+
+	/* Calculate PMP config/addr register and shift */
+#ifdef CONFIG_64BIT
+	pmpcfg_csr = CSR_PMPCFG0 + (index >> 4);
+	shift = (index & 0x0007) << 3;
+#else
+	pmpcfg_csr = CSR_PMPCFG0 + (index >> 2);
+	shift = (index & 0x0003) << 3;
+#endif /* CONFIG_64BIT */
+	pmpaddr_csr = CSR_PMPADDR0 + index;
+
+	*cfg_val = (csr_read_enum(pmpcfg_csr) >> shift) & 0xFF;
+	*addr_val = csr_read_enum(pmpaddr_csr) << PMP_SHIFT_ADDR;
+
+	return 0;
+}
+
+void z_riscv_pmp_clear_config(void)
+{
+	for (unsigned int i = 0; i < RISCV_PMP_CFG_NUM; i++)
+		csr_write_enum(CSR_PMPCFG0 + i, 0);
+}
+
+/* Function to help debug */
+void z_riscv_pmp_print(unsigned int index)
+{
+	ulong_t cfg_val;
+	ulong_t addr_val;
+
+	if (pmp_get(index, &cfg_val, &addr_val))
+		return;
+#ifdef CONFIG_64BIT
+	printf("PMP[%d] :\t%02lX %16lX\n", index, cfg_val, addr_val);
+#else
+	printf("PMP[%d] :\t%02lX %08lX\n", index, cfg_val, addr_val);
+#endif /* CONFIG_64BIT */
+}
+
+#if defined(CONFIG_USERSPACE)
+void z_riscv_init_user_accesses(struct k_thread *thread)
+{
+	unsigned char index;
+	unsigned char *uchar_pmpcfg;
+
+	index = 0;
+	uchar_pmpcfg = (unsigned char *) thread->arch.u_pmpcfg;
+
+#ifdef CONFIG_PMP_STACK_GUARD
+	index++;
+#endif /* CONFIG_PMP_STACK_GUARD */
+
+	/* MCU state */
+	thread->arch.u_pmpaddr[index] = TO_PMP_ADDR((ulong_t) &is_user_mode);
+	uchar_pmpcfg[index++] = PMP_NA4 | PMP_R;
+
+	/* Program and RO data */
+	thread->arch.u_pmpaddr[index] = TO_PMP_ADDR(0x20000000);
```

Comment:
We don't want to hard code the location of Program/RO memory here. Beside the fact that this could be wrong or overly permissive on different SOC families, it also can cause problems for SOCs that have both code and R/W memory in RAM. Can this be derived from existing sources like the linker/Kconfig/DT?

Reply:
Is it true. I don't fixed that because @lindemer planned to change it with his patch after this PR.
Is it OK for you ?

Reply:
yeah this ought to come from DTS

question: is this a XIP system? i.e. is text/rodata on flash? if so, consider a single PMP slot, power of two sized, that covers the entire flash region read-only, execute

Reply:
For what it's worth the system I'm testing this on is not XIP so forcing a power of 2 size could be pretty inefficient under that scenario. Perhaps this can be improved at a later date, but for now I've had to patch this in order for it to work.

Reply:
Fixed